NPi.GPIO support for Nano Pi NEO/NEO2 boards

Tungsteno74 opened the pull request:
## Checklist:
  - [x] The code change is tested and works locally.
<del>- [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass** **(how? i haven't tox)**</del>
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) **([pulled to main repo](https://github.com/home-assistant/home-assistant.io/pull/9509))**

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`. <del>**(how? manually? by run it, i receive error to import `script` module)**</del>
  - [x] New files were added to `.coveragerc`. **(how? manually? by script?)**

<del>If the code does not interact with devices: </del>
<del>- [ ] Tests have been added to verify that the new code works. **(how? i don't see any rpi_gpio files in tests folder)**</del>

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

Tungsteno74 committed:
NPi.GPIO support for Nanopi NEO/NEO2 boards

Tungsteno74 committed:
Update binary_sensor.py

Tungsteno74 committed:
Update cover.py

Tungsteno74 committed:
Update switch.py

Tungsteno74 committed:
Update switch.py

Tungsteno74 committed:
Update manifest.json

Tungsteno74 committed:
added initial state

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 commented:
I'm not the owner of the NanoPi.
I still have to sign in the CLA?

Tungsteno74 committed:
Update cover.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update cover.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update cover.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update cover.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update cover.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update __init__.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update binary_sensor.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update switch.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update binary_sensor.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update binary_sensor.py

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented:
Most of your questions are answered in the [developer documentation](https://developers.home-assistant.io/docs/en/development_index.html).

> I'm not the owner of the NanoPi.

So, this integration is not tested?

Tungsteno74 commented:
> Most of your questions are answered in the [developer documentation](https://developers.home-assistant.io/docs/en/development_index.html).
> 
> > I'm not the owner of the NanoPi.
> 
> So, this integration is not tested?

I've tested it on a NanoPi NEO, but i'm not associated with NanoPI team of company.
Plus other NanoPi with H3/H5 chips should be pin mapping compatible, but as i own only one model I can't test the others.

Tungsteno74 committed:
Update manifest.json

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update requirements_all.txt

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 commented:
@fabaff 
After I set up the dev project as described in the [docs](https://developers.home-assistant.io/docs/en/development_environment.html) you mentioned, I have some issue to complete all steps above.

First, after i ran the `python3 -m script.hassfest` command and received this error:
```
(venv) admin@NanopiNEO:~/home-assistant$ python3 -m script.hassfest
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/admin/home-assistant/script/hassfest/__main__.py", line 88, in <module>
    sys.exit(main())
  File "/home/admin/home-assistant/script/hassfest/__main__.py", line 40, in main
    plugin.validate(integrations, config)
  File "/home/admin/home-assistant/script/hassfest/dependencies.py", line 64, in validate
    validate_dependencies(integration)
  File "/home/admin/home-assistant/script/hassfest/dependencies.py", line 41, in validate_dependencies
    r"hass\.components\.(\w+)")
  File "/home/admin/home-assistant/script/hassfest/dependencies.py", line 19, in grep_dir
    for match in pattern.finditer(fil.read_text()):
  File "/usr/lib/python3.6/pathlib.py", line 1197, in read_text
    return f.read()
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 399: ordinal not in range(128)

```
Thus, i decided to remove the codeowner string in the component manifest.

Second, 
https://developers.home-assistant.io/docs/en/development_checklist.html
this not specify how `.coveragerc` is generated, so i decided to add it manually (based on rpi_gpio).

No more problem for me, tell me if works anyway.

Tungsteno74 committed:
Update manifest.json

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
added npi_gpio exclusion from tests

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

Tungsteno74 committed:
Update manifest.json

homeassistant commented:
Hi @Tungsteno74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#24070).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented:
The CLA hasn't been signed for a month. I'll close this PR now. Please open a new PR when you're ready to sign the CLA and continue working on this feature.

Tungsteno74 commented:
> The CLA hasn't been signed for a month. I'll close this PR now. Please open a new PR when you're ready to sign the CLA and continue working on this feature.

hello @MartinHjelmare 
The problem about CLA is that the NanoPi team has not yet replyed on the official forum. 
For this, I don't know if I have the rights to mark the CLA on my behalf.

It's a pity because, the code is ready and working for those models. :(

MartinHjelmare commented:
We don't accept, and won't review, PRs without CLA signed.

Tungsteno74 commented:
[From the CLA page:](https://www.home-assistant.io/blog/2017/01/21/home-assistant-governance/#what-is-a-contributor-license-agreement)

> A Contributor License Agreement, known as a CLA, asks contributors to confirm that:
> 
> Anyone can use your contributions anywhere, for free, forever.
> Your contributions do not infringe on anyone else’s rights.

OK, now i understood the CLA better. I think that i can sign-in without NanoPi team approval, as the code that i share here was written by me.

Tungsteno74 committed:
updated npi.gpio version

Tungsteno74 committed:
updated version to latest

MartinHjelmare commented on homeassistant/components/npi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,89 @@
+    invert_logic = config.get(CONF_INVERT_LOGIC)
+
+    binary_sensors = []
+    ports = config.get('ports')
```
Use `CONF_PORTS`.

Please use `dict[key]` for required config keys and keys with default config schema values.

Tungsteno74 replied:
> Please use `dict[key]` for required config keys and keys with default config schema values.

...sorry, but i do not undertand well how to apply this.
Could you give me a code example? 
thanks

MartinHjelmare replied:
```py
ports = config[CONF_PORTS]
```

Tungsteno74 replied:
Done in 8d71d7b485e50f5abf46a3fafd0739c772bbe201 dc14a6aee372fd7a9e87f9af7dcfde1747ccf640 ffe24b2e44277423971addad5d6b10e633091fc5

Tungsteno74 replied:
tested and works

MartinHjelmare commented on homeassistant/components/npi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,89 @@
+
+import voluptuous as vol
+
+from homeassistant.components import npi_gpio
```
Please make all intra package imports relative.
```py
from . import setup_input, ...
```

Tungsteno74 replied:
Done in commits 06ee5e246ec118117fac5d7c52041bfc53df0d38 6d725770b585615365e65b89f168e4b8dc5b29cf 9b427283e223f70bc1296fc7dbbb0e3b1e4cffd6 47ab26acf3610677c85f2b8da70149016f5d994f

Tungsteno74 replied:
tested and works

MartinHjelmare commented on homeassistant/components/npi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,89 @@
+        self._invert_logic = invert_logic
+        self._state = None
+
+        npi_gpio.setup_input(self._port, self._pull_mode)
```
We want to avoid side effects in init method. Please move this call out of the class or into a dedicated setup method.

Tungsteno74 replied:
I'm not able to fix that. I can't find a reference code or example to do that for that kind of component. Both Arduino and Beagle Bone GPIO's components (that i used as reference) use this setup.

MartinHjelmare replied:
Add a method `setup` in the entity. Move this call to that method. Call the method on each entity after creating the entity before adding the entities.

MartinHjelmare commented on homeassistant/components/npi_gpio/binary_sensor.py:
```diff
@@ -0,0 +1,89 @@
+            self._state = npi_gpio.read_input(self._port)
+            self.schedule_update_ha_state()
+
+        npi_gpio.edge_detect(self._port, read_gpio, self._bouncetime)
```
State update callbacks should be registered in the entity coroutine method `async_added_to_hass`. Please move this to that method and schedule the call on the executor thread pool with `hass.async_add_executor_job`.

MartinHjelmare commented on homeassistant/components/npi_gpio/cover.py:
```diff
@@ -0,0 +1,112 @@
+    def _trigger(self):
+        """Trigger the cover."""
+        npi_gpio.write_output(self._relay_port, 1 if self._invert_relay else 0)
+        sleep(self._relay_time)
```
We're not allowed to sleep in the executor thread pool. We would risk exhausting it.

Either remove this or switch to the home assistant async api to be able to use `asyncio.sleep`. Async sleep can be done without penalty.

```py
async def async_close_cover(self):
    if not self.is_closed:
        await self._trigger()

async def _trigger(self):
    await hass.async_add_executor_job(
        npi_gpio.write_output,
        self._relay_port,
        1 if self._invert_relay else 0)
    await asyncio.sleep(self._relay_time)
    await hass.async_add_executor_job(
        npi_gpio.write_output,
        self._relay_port,
        0 if self._invert_relay else 1)
```

Tungsteno74 replied:
Sleep is necessary to avoid the failure of one of the two commands run sequentially. I'm trying to change this part with async support.

Tungsteno74 replied:
tested and works

Tungsteno74 committed:
Update binary_sensor.py

discussion https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303122989

Tungsteno74 committed:
Update cover.py

discussion https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303122989

Tungsteno74 committed:
Update switch.py

discussion https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303122989

Tungsteno74 committed:
Update binary_sensor.py

discussion https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303122989

Tungsteno74 committed:
Update binary_sensor.py

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303122380

Tungsteno74 committed:
changed triggers for cover as async fuctions

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303126958

Tungsteno74 committed:
updated entity attribute hass with self.hass

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303126958

Tungsteno74 committed:
changed config.get(key) to config[key]

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303194196

Tungsteno74 committed:
changed config.get[key] to config[key]

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303194196

Tungsteno74 committed:
changed config.get[key] to config[key]

comment https://github.com/home-assistant/home-assistant/pull/24070#discussion_r303194196

Tungsteno74 committed:
Merge pull request #2 from Tungsteno74/npigpio_component-dev-fixes

Npigpio component dev fixes

MartinHjelmare commented on homeassistant/components/npi_gpio/cover.py:
```diff
@@ -1,12 +1,12 @@
 """Support for controlling a Nano Pi cover."""
 import logging
-from time import sleep
+from asyncio import sleep as asleep
```
Please `import asyncio` and use `asyncio.sleep` in the code.

MartinHjelmare commented:
@Tungsteno74 are you planning to finish here?

MartinHjelmare commented:
Closing due to inactivity.
